django/common/kafka/consumer.py

```diff
@@ -56,34 +51,33 @@ def _generate_config(self):
         }
         return config
 
-    def consume_event(self):
-        """
-        Consume event in an infinite loop
-        :return:
-        """
+    @contextmanager
+    def subscribe(self):
+        self.consumer.subscribe(self.topics)
+        yield Subscription(self.consumer)
+        self.consumer.close()
+
+
+class Subscription:
+    def __init__(self, consumer) -> None:
+        self.consumer = consumer
+
+    def __call__(self) -> dict:
+        """Polls until a message's available for delivery"""
         while True:
-            # Deserialize Event
             msg = self.consumer.poll(timeout=1.0)
-
-            # Process Event or Raise Error
             if msg is None:
                 continue
-            if msg.error():
-                self.manage_error(msg)
-            else:
-                # Proper message
-                self.process_event(msg)
+            if error := msg.error():
+                logger.debug(error.str())
+                self.handle_error(error)
+                continue
+            data = json.loads(msg.value())
+            return data
 
-    def run_consumer(self):
-        """
-        Create consumer, assign topics, consume and process events
-        :return:
-        """
-        self.consumer.subscribe(self.topics)
-        try:
-            self.consume_event()
-        except (KafkaException, KafkaError):
-            raise
-        finally:
-            # Close down consumer to commit final offsets.
-            self.consumer.close()
+    def handle_error(self, error: KafkaError):
+        """Handles a KafkaError"""
+        if error.retriable() or error.code() == KafkaError.UNKNOWN_TOPIC_OR_PART:
+            time.sleep(1)
```

Comment:
Can we add a comment to explain why a sleep is needed here ?

Reply:
`UNKNOWN_TOPIC_OR_PART` means that there is no such topic available, i.e. nothing to do bc noone is using the ETL. I expect that we'll be in this situation 99% of the time. Sleeping here to save electricity ;)